src/meta-srv/src/election/postgres.rs

```diff
@@ -343,6 +457,148 @@ impl PgElection {
+                        // Case 1.2: I'm still the leader but the lease expired. Step down and re-init campaign.
+                        (true, false) => {
+                            warn!("Leader lease expired, now stepping down.");
+                            self.step_down().await?;
+                        }
+                        // Rarely happens:
+                        // Case 1.3: I'm not the leader. This happens when I failed to step down in the previous term and the lock released
+                        //           after the session expired. And then I reconnect and find a gap that others just released the lock.
+                        //           I should re-step-down and re-init campaign.
+                        (false, _) => {
+                            warn!("Leader lease not found, but still hold the lock. Now stepping down.");
+                            self.step_down().await?;
+                        }
+                    }
+                }
+                // Case 1.4: Same as case 1.3, just re-step-down and re-init campaign.
+                None => {
+                    warn!("Leader lease not found, but still hold the lock. Now stepping down.");
+                    self.step_down().await?;
+                }
+            }
+        // Case 2: I'm not the leader in the previous term.
+        } else {
+            self.elected().await?;
+        }
+        Ok(())
+    }
+
+    async fn follower_action(&self) -> Result<()> {
+        let key = self.election_key();
+        // Case 1: I think I'm the leader in the previous term but failed to acquire the lock. I just step down.
+        if self.is_leader() {
+            self.step_down_without_lock().await?;
+        }
+        let (_, expire_time, current, _) = self
+            .get_value_with_lease(&key, false)
+            .await?
+            .context(NoLeaderSnafu)?;
+        // Case 2: Current leader exists and the lease expired. Raise an error to re-init campaign.
+        //         If the leader failed to renew the lease, its session will expire and the lock will be released.
+        ensure!(expire_time > current, NoLeaderSnafu);
+        // Case 3: All checks passed, just return.
+        Ok(())
+    }
+
+    /// Step down the leader. The leader should delete the key and notify the leader watcher.
+    /// Do not check if the deletion is successful, since the key may be deleted by other followers.
+    /// Caution: Should only step down while holding the advisory lock.
+    async fn step_down(&self) -> Result<()> {
+        let key = self.election_key();
+        let leader_key = PgLeaderKey {
+            name: self.leader_value.clone().into_bytes(),
+            key: key.clone().into_bytes(),
+            ..Default::default()
+        };
+        if self
+            .is_leader
+            .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed)
+            .is_ok()
+        {
+            self.delete_value(&key).await?;
```

Comment:
Is it possible that this operation mistakenly deletes the key written by the new leader?

Reply:
The deletion will only happens when the metasrv holds an advisory lock. If a new leader is elected, it should have held the lock to prevent others from the deletion.

Reply:
Got it